pgjdbc/src/main/java/org/postgresql/jdbc/TimestampUtils.java

```diff
@@ -1514,27 +1520,41 @@ private static long toPgSecs(long secs) {
   /**
    * Converts the SQL Date to binary representation for {@link Oid#DATE}.
    *
-   * @param tz The timezone used.
+   * @param tz The timezone used. If <code>null</code>, then no timezone offset will be applied to the date.
    * @param bytes The binary encoded date value.
    * @param value value
    * @throws PSQLException If binary format could not be parsed.
    */
   public void toBinDate(@Nullable TimeZone tz, byte[] bytes, Date value) throws PSQLException {
     long millis = value.getTime();
-
+    boolean isBce = false;
     if (tz == null) {
       tz = getDefaultTz();
     }
-    // It "getOffset" is UNTESTED
-    // See org.postgresql.jdbc.AbstractJdbc2Statement.setDate(int, java.sql.Date,
-    // java.util.Calendar)
-    // The problem is we typically do not know for sure what is the exact required date/timestamp
-    // type
-    // Thus pgjdbc sticks to text transfer.
-    millis += tz.getOffset(millis);
+    if (millis < GREGORIAN_CUTOVER) {
+      BaseCalendar calendar = getJulianCalendar();
+      CalendarDate calendarDate = calendar.getCalendarDate(millis, tz);
+      isBce = calendarDate.getEra().getAbbreviation().equals("B.C.E.");
```

Comment:
I disagree with that check. It should not be string based.

Reply:
ugh, ya, that seems like a bad idea

Reply:
I would use code like: `IsoEra.BCE.getValue()` that returns the integer value of BC. We use that at other places quite often.

In general the code hiere tries to fix a problem which is caused by the broken `java.sql.Date#toLocalDate()`. Doing this is risky, see my other comments about the "one year off" problem. If you print the date with toString() it returns a negative year (but one off).

Reply:
Oh thats on legacy java.util.Calendars.... Why is this PR using "sun." packages? That should never be done, as this is subject to change! Sorry that should never be included in pgjdbc (see import statements).

Reply:
To be honest I am only using the PR for tests. Perhaps we should create another PR which implements better tests and a proposed fix ?